datafusion/expr/src/udf.rs

```diff
@@ -84,9 +94,60 @@ impl ScalarUDF {
         }
     }
 
+    /// Create a new `ScalarUDF` from a `FuncImpl`
+    pub fn new_from_impl(
+        arc_fun: Arc<dyn FunctionImplementation + Send + Sync>,
+    ) -> ScalarUDF {
+        let captured_self = arc_fun.clone();
+        let return_type: ReturnTypeFunction = Arc::new(move |arg_types| {
+            let return_type = captured_self.return_type(arg_types)?;
+            Ok(Arc::new(return_type))
+        });
+
+        let captured_self = arc_fun.clone();
+        let func: ScalarFunctionImplementation =
+            Arc::new(move |args| captured_self.invoke(args));
+
+        ScalarUDF::new(arc_fun.name(), arc_fun.signature(), &return_type, &func)
+    }
+
     /// creates a logical expression with a call of the UDF
     /// This utility allows using the UDF without requiring access to the registry.
     pub fn call(&self, args: Vec<Expr>) -> Expr {
         Expr::ScalarUDF(crate::expr::ScalarUDF::new(Arc::new(self.clone()), args))
     }
+
+    /// Returns this function's name
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+    /// Returns this function's signature
+    pub fn signature(&self) -> &Signature {
+        &self.signature
+    }
+    /// return the return type of this function given the types of the arguments
+    pub fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        // Old API returns an Arc of the datatype for some reason
+        let res = (self.return_type)(args)?;
+        Ok(res.as_ref().clone())
+    }
+    /// return the implementation of this function
+    pub fn fun(&self) -> &ScalarFunctionImplementation {
+        &self.fun
+    }
+}
+
+/// Convenience trait for implementing ScalarUDF. See [`ScalarUDF::from_impl()`]
```

Comment:
This echo's the trait that @2010YOUY01  proposed in ) https://github.com/apache/arrow-datafusion/pull/7752, but does so in a way that is backwards compatible (makes a ScalarUDF out of the trait, to retain backwards compatibly)